src/kernel_histogram.cpp

```diff
@@ -40,7 +40,11 @@ kernel_histogram(const unsigned *dim,
   const size_t size_out_y = dim[1];
   const int radius = dim[2];
   const size_t kernel_size = 2 * radius + 1;
-  const size_t num_threads = dim[3];
+
+  size_t num_threads = dim[3];
+  if (num_threads == 0)
+    num_threads = thread::hardware_concurrency();
+
   vector<thread> list_threads(num_threads);
   const size_t block_size_y = round(size_out_y / (float)num_threads);

```

Comment:
- style: better use c++ `float(...)` conversion instead of C `(float)...`
- overall, indices should never get rounded or divided by floats. I sent a suggestion on mattermost

Also, there could be a function (higher-order probably) that abstracts the parallel processing